### PR TITLE
review README

### PR DESCRIPTION
**Pull request by meghalithic**
issue #40

updated the readme based on discussions at the workflow meeting and comments on issue 40 and PR #36

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,15 +1,20 @@
 # Morphology-analysis
-Extract morphological characteristics from image of fish trait segmentation.
 
-The goals of the tool is to extract presence-absence table, measurements and landmarks of fish from the segmented fish image produced by M. Maruf.
-It provides a framework for creating modularized tools by using classes and jupyter notebook to visualize and test functionalities. 
-The tool is automatically containerized when a new release is published. It provides a validated version for easy integration into the [BGNN_Snakemake](https://github.com/hdr-bgnn/BGNN_Snakemake).
-This tool can me made more generalizable but it has been developed for the [Minnows Project](https://github.com/hdr-bgnn/minnowTraits).
+The primary goal of this repository is to produce a presence table (default), create landmarks (optional), visualize landmarks (otpional, and extract measurements (optional) from a segmented fish image produced by M. Maruf (see [BGNN-trait-segmentation](https://github.com/hdr-bgnn/BGNN-trait-segmentation)).
```
Comment (hlapp) [inline README.md]:
```suggestion
The primary goal of this repository is to produce a presence table (default), create landmarks (optional), visualize landmarks (optional), and extract measurements (optional) from a segmented fish image (see [BGNN-trait-segmentation](https://github.com/hdr-bgnn/BGNN-trait-segmentation)).
```

File: README.md [hunk 2]
```diff
@@ -1,15 +1,20 @@
...
 
+Secondarily, this repository provides a framework for creating modularized tools by using classes.
 
-## 1- Segmented image .png description
+Finally, this repository also provides a way to visualize outputs of the tools and test functionalities using a jupyter notebook. 
 
-The segmented image input looks like image below, with traits color coded and identified by "blobs". The segmentation model uses [M. Maruf's segmentation code](https://github.com/hdr-bgnn/BGNN-trait-segmentation/blob/main/Segment_mini/scripts/segmentation_main.py), and is based on a Convolutional Neural Network (CNN; more specifically unet). You can find more information on the [BGNN-trait-segmentation repository](https://github.com/hdr-bgnn/BGNN-trait-segmentation).  The output is 11 classes (traits) that are color coded. We are only using 9 of them, and are excluding alt_fin_ray and caudal_fin_ray.
+This repository is automatically containerized when a new release is published. This allows for easy integration into the [BGNN_Snakemake](https://github.com/hdr-bgnn/BGNN_Snakemake).
```
Comment (hlapp) [inline README.md]:
```suggestion
A Docker container image of this codebase and its dependencies is automatically built and published each time a new release is published. The Docker container isolates dependencies and facilitates reproducible and version-tagged integration into Imageomics workflows (see for example the [BGNN core workflow](https://github.com/hdr-bgnn/BGNN_Snakemake)).
```

File: README.md [hunk 3]
```diff
@@ -1,15 +1,20 @@
...
+
+This tool can me made more generalizable, but was originally developed for the [Minnows Project](https://github.com/hdr-bgnn/Minnow_Segmented_Traits).
```
Comment (hlapp) [inline README.md]:
```suggestion
This tool was originally developed for the [Minnows Project](https://github.com/hdr-bgnn/Minnow_Segmented_Traits).
```

File: README.md [hunk 4]
```diff
@@ -1,15 +1,20 @@
...
+
+## 1- Input: Segmented Image
+
+The input is the output of the segmentation model, which is a file named basename_segmented.png, where "basename" is the file name for the original image. The segmentation model uses [M. Maruf's segmentation code](https://github.com/hdr-bgnn/BGNN-trait-segmentation/blob/main/Segment_mini/scripts/segmentation_main.py) and is based on a Convolutional Neural Network (CNN; more specifically unet). More information can be found on the [BGNN-trait-segmentation repository](https://github.com/hdr-bgnn/BGNN-trait-segmentation).
```
Comment (hlapp) [inline README.md]:
```suggestion
The input is the output of the segmentation model, which is a file named basename_segmented.png, where "basename" is the file name for the original image. The segmentation model is based on a Convolutional Neural Network (CNN; more specifically unet). More information can be found In the [BGNN-trait-segmentation repository](https://github.com/hdr-bgnn/BGNN-trait-segmentation), in particular see the [segmentation code there](https://github.com/hdr-bgnn/BGNN-trait-segmentation/blob/main/Segment_mini/scripts/segmentation_main.py).
```

File: README.md [hunk 5]
```diff
@@ -1,15 +1,20 @@
...
+
+The segmented image looks like image below, with traits color-coded and visualized by "blobs".  There are 11 trait classes corresponding to the annotated traits. Here, only 9 are used (alt_fin_ray and caudal_fin_ray are excluded).
```
Comment (hlapp) [inline README.md]:
```suggestion
A segmented image looks like the image below, with traits color-coded and visualized by "blobs".  There are 11 trait classes corresponding to the annotated traits. Here, only 9 are used (alt_fin_ray and caudal_fin_ray are excluded).
```

File: README.md [hunk 6]
```diff
@@ -30,27 +35,38 @@ When you export this image in python using pillow library (PIL.Image.open(file_n
 * 'trunk': [0, 124, 124]
 
 
-## 2- Presence Absence of Traits
+## 2- Default Tool: Presence
+
+The input of the presence class, or tool, is a semgemented image named "basename_segmented.png" generated by [Virginia Tech](https://github.com/hdr-bgnn/BGNN-trait-segmentation).
```
Comment (hlapp) [inline README.md]:
```suggestion
The input of the presence tool, is a segmented image named "basename_segmented.png" (as generated by the [BGNN-trait-segmentation](https://github.com/hdr-bgnn/BGNN-trait-segmentation) code).
```

File: README.md [hunk 7]
```diff
@@ -30,27 +35,38 @@ When you export this image in python using pillow library (PIL.Image.open(file_n
...
+
+The output of the presence tool is a presence table named "basename_presence.json" for each image.
+
+To check for the presence of traits, the tool:
 
-The approach that we use for checking presence and absence of traits is the following:
+  1. Isolates an individual trait (e.g., isolates the dorsal_fin)
+  2. Counts the number of blobs for that trait
+  3. Calculates the percentage of the area of the largest blob as a proportion of the area of the total blobs for that trait
```
Comment (hlapp) [inline README.md:48]: So if there's only one blob for a trait, the result is automatically 100%? Or is this meant to say something different?

File: README.md [hunk 8]
```diff
@@ -30,27 +35,38 @@ When you export this image in python using pillow library (PIL.Image.open(file_n
...
 
-  1. Isolate indivual traits (e.g., isolate the dorsal_fin)
-  2. Count number of blobs
-  3. Calculate percentage of large blob
 
+## 3- Optional Tools: Landmark, Visualize, Metadata, and Measure
 
-## 3- Other Functions
+The input for the measure tool is a metadata file named "basename_metadata.json" generated by [Drexel](https://github.com/hdr-bgnn/drexel_metadata/tree/Thibault/gen_metadata_mini).
```
Comment (hlapp) [inline README.md]: So this refers to a branch in the drexel_metadata repository, which is almost certainly not what we want to do. Hence, the correct (in the sense of more canonical) answer is presumably different?

File: README.md [hunk 9]
```diff
@@ -30,27 +35,38 @@ When you export this image in python using pillow library (PIL.Image.open(file_n
...
 
-The approach that we use for creating landmarks and extracting measurements is the following: 
+To create landmarks, visualizations, and extract measurements, the tool:
+
+  1. Isolates an individual trait (e.g., isolates the dorsal_fin)
+  2. Removes any small blobs of a trait and fills in gaps between blobs of a trait
```
Comment (hlapp) [inline README.md]: This is a little confusing. Are the gaps only filled in between "major" blobs of a trait, and what decides which blurb of a trait is a small one that gets removed, and which blurb is instead not only kept but also has the "gap" filled in?

File: README.md [hunk 10]
```diff
@@ -77,9 +93,30 @@ If you had more features in the class and codes to extract landmarks or measurem
...
+### Visualize
+
+This tool saves the segmented fish image with landmarks. 
+
+Here is an [example](Test_Data/INHS_FISH_000742_image_lm.png) of a visualization of the landmarks on the segmented image.
+
+
+### Metadata
+
+This is an input file. The file should be generated by [drexel](https://github.com/hdr-bgnn/drexel_metadata/tree/Thibault/gen_metadata_mini) and named basename_metadata.json.
```
Comment (hlapp) [inline README.md:108]: Again, we are referring to a branch in a repository, almost certainly not what we want to do. Presumably there's a better (more canonical) reference?

File: README.md [hunk 11]
```diff
@@ -77,9 +93,30 @@ If you had more features in the class and codes to extract landmarks or measurem
...
+This file contains information about the rule and scale (pixels per cm) to be used for the **Measure** tool.
```
Comment (hlapp) [inline README.md:109]:
```suggestion
This file contains information about the ruler and scale (pixels per cm) in the image, to be used by the **Measure** tool. 
```

File: README.md [hunk 12]
```diff
@@ -77,9 +93,30 @@ If you had more features in the class and codes to extract landmarks or measurem
...
+
+Here is an [example](Test_Data/INHS_FISH_000742_image_lm.png) of a visualization of the landmarks on the segmented image.
+
+
+### Metadata
+
+This is an input file. The file should be generated by [drexel](https://github.com/hdr-bgnn/drexel_metadata/tree/Thibault/gen_metadata_mini) and named basename_metadata.json.
+This file contains information about the rule and scale (pixels per cm) to be used for the **Measure** tool. 
+
+The scale function within this tool outputs if a ruler is found, the scale in pixels, and the unit (cm).
```
Comment (hlapp) [inline README.md]:
```suggestion
The scale function within this tool outputs the scale in pixels, and the unit (cm), provided a ruler is found,
```

Comment (hlapp) [reply README.md]: Makes me wonder what it does if a ruler is not found.

File: README.md [hunk 13]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
   distance  |  eye diameter using landmarks  |  ED_bbox  |  length across the eye along the anterior-posterior (left-right) axis (distance between the left-right sides of a bounding box around the eye)  
   angle  |  fish angle using landmarks  |  Fa_lm  |  angle of the tilt of the fish from horizontal (angle between the SL_lm and and the horizontal line of the image)  
   angle  |  fish angle using PCA  |  Fa_pca  |  angle of the tilt of the fish from horizontal (angle between the pca through the midline of the fish mask and the horizontal line of the image)  
+
   
-#### method of measurement extraction
+#### Method of Measurement Extraction
+
+Each method of measurement extracted is a separate class, which adds more flexibility. These are described in [Trait_class](Scripts/Traits_class.py)
```
Comment (hlapp) [inline README.md]:
```suggestion
Each method of measurement extraction is in a separate Python class. These are described in [Trait_class](Scripts/Traits_class.py)
```

File: README.md [hunk 14]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
 
 _landmarks_
 
-These measurement trait classes functions have the suffix "_lm"_ to denote the method of extraction. 
-The lengths (in pixels) are calculated as the distance between two landmarks (described in the "Definition" column of the trait description csvs).
 
-_bbox (bounding box)_
-These trait classes functions have the suffix "_bbox"_ to denote the method of extraction.
-The lengths (in pixels) are calculated as the distance of a perpindicular line between the edges (either vertical or horizontal) of the bbox.
+These measurement trait classes functions have the suffix *"_lm"* to denote the method of extraction.
```
Comment (hlapp) [inline README.md]:
```suggestion
These trait measurement class functions have the suffix *"_lm"* to denote the method of extraction. 
```

File: README.md [hunk 15]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
-These trait classes functions have the suffix "_bbox"_ to denote the method of extraction.
-The lengths (in pixels) are calculated as the distance of a perpindicular line between the edges (either vertical or horizontal) of the bbox.
+These measurement trait classes functions have the suffix *"_lm"* to denote the method of extraction. 
+The lengths (in pixels) are calculated as the distance between two landmarks (described in the "Definition" column of the trait description csv).
+
+
+_bounding box (bbox)_
+
+
+These trait classes functions have the suffix *"_bbox"* to denote the method of extraction.
```
Comment (hlapp) [inline README.md]:
```suggestion
These trait class functions have the suffix *"_bbox"* to denote the method of extraction.
```

File: README.md [hunk 16]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
+The lengths (in pixels) are calculated as the distance of a perpendicular line between the edges (either vertical or horizontal) of the bounding box (bbox).
 
-_mask_
-These trait classes have the suffix "_m"_ to denote the method of extraction.
 
 #### Areas
 
-Areas are calculated as the total pixels in the mask of a trait (e.g., head area is the area of the mask of the head). These are described in the "Definition" column of the <a href="https://github.com/hdr-bgnn/Morphology-analysis/blob/main/Traits_description/Minnow_Measurements_v1.csv">Minnow_Measurements_v1.csv</a>.
+
+Areas are calculated as the total pixels in the mask of a trait (e.g., head area is the area of the mask of the head). These are described in the "Definition" column of the [Minnow_Measurements_v1.csv](https://github.com/hdr-bgnn/Morphology-analysis/blob/main/Traits_description/Minnow_Measurements_v1.csv).
```
Comment (hlapp) [inline README.md]:
```suggestion
Areas are calculated as the total pixels in the mask of a trait (e.g., head area is the area of the mask of the head). These are described in the "Definition" column of [Minnow_Measurements_v1.csv](Traits_description/Minnow_Measurements_v1.csv).
```

File: README.md [hunk 17]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
 
+The outputs for the tools are:
+
+* Presence: basename_presence.json (**required**)
+* Landmark: basename_landmark.json (*optional*)
+* Visualize: basename_lm_image.png (*optional*)
+* Measure: basename_morphology.json (*optional*)
+
+
+The main arguments to run this repository are:
```
Comment (hlapp) [inline README.md]:
```suggestion
The main arguments for running `Morphology_main.py` are:
```

File: README.md [hunk 18]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
+--visualize VISUALIZE.png
+```
+
+*measure*
+```
+--measure  MEASURE.json
+```
+
+
+Here is an example using [Test_Data](https://github.com/hdr-bgnn/Morphology-analysis/Test_Data):
```
Comment (hlapp) [inline README.md]:
```suggestion
Here is an example using [test data](Test_Data): 
```

File: README.md [hunk 19]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
+```
+
+*measure*
+```
+--measure  MEASURE.json
+```
+
+
+Here is an example using [Test_Data](https://github.com/hdr-bgnn/Morphology-analysis/Test_Data): 
 ```
```
Comment (hlapp) [inline README.md]:
```suggestion
```sh
```

File: README.md [hunk 20]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
- + landmark.json : **Optional**. Save the landmark dictionnary with the filename provided. dictionnary, key = landmark label, value = calculated value [example](Test_Data/INHS_FISH_000742_landmark.json)
-  + image_lm.png : **Optional**. Save the visualization image for landmarks with the filename provided. [example](Test_Data/INHS_FISH_000742_image_lm.png)
-  + input_file.png : **Positional require**. Segmented fish image generated by [Maruf code](https://github.com/hdr-bgnn/BGNN-trait-segmentation/tree/main/Segment_mini), [example](Test_Data/INHS_FISH_000742_segmented.png)
- + ouput_presence.json : **Positional require**. Save the presence-absence dictionnary with the filename provided [example](Test_Data/INHS_FISH_000742_presence.json)
 
-## 5- Container, usage and release
+If no arguments are given, an error message will say "missing two positional arguments", which are the input file and the output file. Use "-h" to pull up the help file with the full list of arguments. 
+
+
+## 5- Version control
```
Comment (hlapp) [inline README.md]:
```suggestion
## 5- Automatic Docker container build and version tags
```

File: README.md [hunk 21]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
+
+The repository is containerized each release and is stored on the GitHub registry. The containerized version is called to run the main script, [Morphology_main.py](Scripts/Morphology_main.py).
```
Comment (hlapp) [inline README.md]:
```suggestion
Upon publishing a new release, a Docker container image is automatically built from the release and published on the GitHub container and package registry. The published image is tagged with major, major.minor, and major.minor.patch versions corresponding to the release.
```

File: README.md [hunk 22]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
 
-## 5- Container, usage and release
+If no arguments are given, an error message will say "missing two positional arguments", which are the input file and the output file. Use "-h" to pull up the help file with the full list of arguments. 
+
+
+## 5- Version control
+
+The repository is containerized each release and is stored on the GitHub registry. The containerized version is called to run the main script, [Morphology_main.py](Scripts/Morphology_main.py). 
+
+The workflow to build the container is defined [here](.github/workflows/Deploy_Morpholgy.yml).
```
Comment (hlapp) [inline README.md]:
```suggestion
The workflow  to build the container is defined as a GitHub action [here](.github/workflows/Deploy_Morpholgy.yml).   
```

File: README.md [hunk 23]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
+
+The Dockerfile definition is [here](Dockerfile).
+
+Pull the latest image: 
+```
+docker pull ghcr.io/hdr-bgnn/morphology-analysis/morphology:latest 
+#singularity pull docker://ghcr.io/hdr-bgnn/morphology-analysis/morphology:latest
+```
+
+Run the container:
```
Comment (hlapp) [inline README.md]:
```suggestion
Run the container (assuming an HPC environment that supports singularity but not docker): 
```

File: README.md [hunk 24]
```diff
@@ -101,140 +138,185 @@ If you had more features in the class and codes to extract landmarks or measurem
...
+
+## 5- Version control
+
+The repository is containerized each release and is stored on the GitHub registry. The containerized version is called to run the main script, [Morphology_main.py](Scripts/Morphology_main.py). 
+
+The workflow to build the container is defined [here](.github/workflows/Deploy_Morpholgy.yml).   
+
+The Dockerfile definition is [here](Dockerfile).
+
+Pull the latest image:
```
Comment (hlapp) [inline README.md]:
```suggestion
Pull the latest image (in an HPC environment, docker is typically not supported but singularity is): 
```

**Review by hlapp — CHANGES_REQUESTED** [review]
I think section 6 needs to be overhauled (it should not be specific to the OSC, but probably focus on generally necessary steps, such as creating conda environment; where to go as work directory; etc). Perhaps @johnbradley can take a look?

**Comment by meghalithic** [comment]
@hlapp @johnbradley section 6 was more of notes for myself. It could probably be removed?

**Comment by thibaulttabarin** [comment]
LGTM, one thing though I would insist putting somewhere in the usage section this line :
Morphology_main.py [-h] [--metadata METADATA] [--morphology MORPHOLOGY] [--landmark LANDMARK] [--lm_image LM_IMAGE] input_image output_presence

@hlapp @johnbradley What do you think?

**Comment by meghalithic** [comment]
@thibaulttabarin I broke it out into separate sections. Is it better to have it all together?

**Comment by hlapp** [comment]
> putting somewhere in the usage section this line :
> Morphology_main.py [-h] [--metadata METADATA] [--morphology MORPHOLOGY] [--landmark LANDMARK] [--lm_image LM_IMAGE] input_image output_presence

I think that makes sense as a concise summary of the available options, perhaps best preceding (right before?) the documentation of each individual option.

**Review by thibaulttabarin — APPROVED** [review]
Looks good

File: README.md (hunk 7 above)
Comment (thibaulttabarin) [reply README.md:48]: this is correct, if only one blob <=> 100%

File: README.md (hunk 9 above)
Comment (thibaulttabarin) [reply README.md]: precision which is explained in the code itself and can be understand better using the jupyter notebook by exploring/playing the different functions of the class... 
blob means detach entity, so 2 blobs means to distinct islands (clusters) of pixel.  For "gap" in the description of the code I believe I use the term "hole" which means if you have a blob that present a hole inside (cluster of False pixel or 0 pixel inside the cluster of True pixel), typically the "head" cluster will have a hole corresponding to the eye.

File: README.md (hunk 10 above)
Comment (meghalithic) [reply README.md:108]: Changed

**Comment by meghalithic** [comment]
@hlapp when I made edits it got rid of your comments...how do I see those?

**Comment by hlapp** [comment]
> when I made edits it got rid of your comments...how do I see those?

Just click on the "Outdated" marker, and it should expand to the original comment or edit suggestion.

File: README.md (hunk 12 above)
Comment (meghalithic) [reply README.md]: I added information about that

File: README.md (hunk 8 above)
Comment (meghalithic) [reply README.md]: I put in the link to just the repo. Sorry about that!

File: README.md (hunk 7 above)
Comment (meghalithic) [reply README.md:48]: I added an example. I can add it as an example output if that's better.

File: README.md (hunk 9 above)
Comment (meghalithic) [reply README.md]: I tried to clarify

**Comment by johnbradley** [comment]
> @hlapp @johnbradley section 6 was more of notes for myself. It could probably be removed?

@megbalk I updated section 6 to be more generic, but feel free to remove if you don't feel these instructions are necessary.

**Review by johnbradley — APPROVED** [review]
LGTM

**Comment by meghalithic** [comment]
@hlapp how does @johnbradley change look?

**Review by hlapp — APPROVED** [review]
Looks all good to me now!

**Pull request merged**